ostd/src/mm/page_table/node/mod.rs

```diff
@@ -152,135 +123,85 @@ impl<E: PageTableEntryTrait, C: PagingConstsTrait> RawPageTableNode<E, C> {
     pub(super) unsafe fn first_activate(&self) {
         use crate::{arch::mm::activate_page_table, mm::CachePolicy};
 
-        self.inc_ref_count();
-
-        activate_page_table(self.raw, CachePolicy::Writeback);
+        activate_page_table(self.clone().into_raw(), CachePolicy::Writeback);
     }
+}
 
-    fn inc_ref_count(&self) {
-        // SAFETY: We have a reference count to the page and can safely increase the reference
-        // count by one more.
-        unsafe {
-            inc_frame_ref_count(self.paddr());
-        }
-    }
+/// A reference to a page table node.
+pub(super) type PageTableNodeRef<'a, E, C> = FrameRef<'a, PageTablePageMeta<E, C>>;
 
-    /// Restores the handle from the physical address and level.
+impl<'a, E: PageTableEntryTrait, C: PagingConstsTrait> PageTableNodeRef<'a, E, C> {
+    /// Locks the page table node.
     ///
-    /// # Safety
-    ///
-    /// The caller must ensure that the physical address is valid and points to
-    /// a forgotten page table node. A forgotten page table node can only be
-    /// restored once. The level must match the level of the page table node.
-    pub(super) unsafe fn from_raw_parts(paddr: Paddr, level: PagingLevel) -> Self {
-        Self {
-            raw: paddr,
-            level,
-            _phantom: PhantomData,
+    /// An atomic mode guard is required to
+    ///  1. prevent deadlocks;
+    ///  2. provide a lifetime (`'rcu`) that the nodes are guaranteed to outlive.
+    pub(super) fn lock<'rcu>(self, _guard: &'rcu dyn InAtomicMode) -> PageTableGuard<'rcu, E, C>
+    where
+        'a: 'rcu,
+    {
+        while self
+            .meta()
+            .lock
+            .compare_exchange(0, 1, Ordering::Acquire, Ordering::Relaxed)
+            .is_err()
+        {
+            core::hint::spin_loop();
         }
-    }
-}
 
-impl<E: PageTableEntryTrait, C: PagingConstsTrait> From<RawPageTableNode<E, C>>
-    for Frame<PageTablePageMeta<E, C>>
-{
-    fn from(raw: RawPageTableNode<E, C>) -> Self {
-        let raw = ManuallyDrop::new(raw);
-        // SAFETY: The physical address in the raw handle is valid and we are
-        // transferring the ownership to a new handle. No increment of the reference
-        // count is needed.
-        unsafe { Frame::<PageTablePageMeta<E, C>>::from_raw(raw.paddr()) }
+        PageTableGuard::<'rcu, E, C> {
+            inner: self.clone_ref(),
```

Comment:
```suggestion
            inner: self,
```

Indeed, I'm not seeing any other uses of `clone_ref` (currently)? Should we just remove it?

Reply:
Sure. I'll remove it.